### PR TITLE
Support mlaunch on windows

### PR DESCRIPTION
**Pull request by zhaoyi0113**
Add support on running mlaunch in windows.

### CODE REVIEW

File: mtools/mlaunch/mlaunch.py [hunk 1]
```diff
@@ -668,12 +671,12 @@ def kill(self):
                     sig = getattr(signal, sig)
                 except AttributeError as e:
                     raise SystemExit("can't parse signal '%s', use integer or signal name (SIGxxx)." % sig)
-
         for port in processes:
             # only send signal to matching processes
             if port in matches:
                 p = processes[port]
                 p.send_signal(sig)
+                print " %s on port %i, pid=%i" % (p.name, port, p.pid)
```
Comment (kevinadi) [inline mtools/mlaunch/mlaunch.py]: This line is also printed under the `--verbose` argument.

File: mtools/mlaunch/mlaunch.py [hunk 2]
```diff
@@ -1422,7 +1440,11 @@ def _construct_mongos(self, logpath, port, configdb):
             extra = self._filter_valid_arguments(self.unknown_args, "mongos") + extra
 
         path = self.args['binarypath'] or ''
-        command_str = "%s --logpath %s --port %i --configdb %s --logappend %s %s --fork"%(os.path.join(path, 'mongos'), logpath, port, configdb, auth_param, extra)
+	if os.name == 'nt':
+            newLogPath=logpath.replace('\\', '\\\\')
+            command_str = "start /b %s --logpath %s --port %i --configdb %s --logappend %s %s "%(os.path.join(path, 'mongos'), logpath, port, configdb, auth_param, extra)
+	else:
+            command_str = "%s --logpath %s --port %i --configdb %s --logappend %s %s --fork"%(os.path.join(path, 'mongos'), newLogPath, port, configdb, auth_param, extra)
```
Comment (kevinadi) [inline mtools/mlaunch/mlaunch.py]: I believe `newLogPath` is not defined when the OS is not Windows, resulting in the error `UnboundLocalError: local variable 'newLogPath' referenced before assignment`

File: mtools/mlaunch/mlaunch.py [hunk 3]
```diff
@@ -1422,7 +1440,11 @@ def _construct_mongos(self, logpath, port, configdb):
             extra = self._filter_valid_arguments(self.unknown_args, "mongos") + extra
 
         path = self.args['binarypath'] or ''
-        command_str = "%s --logpath %s --port %i --configdb %s --logappend %s %s --fork"%(os.path.join(path, 'mongos'), logpath, port, configdb, auth_param, extra)
+	if os.name == 'nt':
+            newLogPath=logpath.replace('\\', '\\\\')
+            command_str = "start /b %s --logpath %s --port %i --configdb %s --logappend %s %s "%(os.path.join(path, 'mongos'), logpath, port, configdb, auth_param, extra)
```
Comment (kevinadi) [inline mtools/mlaunch/mlaunch.py]: I assume `newLogPath` was meant to be used in this line?

**Comment by zhaoyi0113** [comment]
Thaks for testing it. Below is my comments regard to your test cases:

1. I haven't tried that paratemers yet. I will let you know after I investigate it.

2. You need to install `openssl` on windows in order to make `--auth` work

3. This command works pretty well for me, below is the output:

```
>mlaunch init --sharded 3 --replicaset --config 3 --mongos 2
launching: start on port 27019
launching: start on port 27020
launching: start on port 27021
launching: start on port 27022
launching: start on port 27023
launching: start on port 27024
launching: start on port 27025
launching: start on port 27026
launching: start on port 27027
launching: config server on port 27028
launching: config server on port 27029
launching: config server on port 27030
replica set 'configRepl' initialized.
replica set 'shard01' initialized.
replica set 'shard02' initialized.
replica set 'shard03' initialized.
launching: start on port 27017
launching: start on port 27018
adding shards. can take up to 30 seconds...
```
4. I will remove the log for this case.

**Comment by zhaoyi0113** [comment]
I see there is an error on travis ci. Could you let me know what os it is running?

File: mtools/mlaunch/mlaunch.py [hunk 4]
```diff
@@ -1422,7 +1439,11 @@ def _construct_mongos(self, logpath, port, configdb):
             extra = self._filter_valid_arguments(self.unknown_args, "mongos") + extra
 
         path = self.args['binarypath'] or ''
-        command_str = "%s --logpath %s --port %i --configdb %s --logappend %s %s --fork"%(os.path.join(path, 'mongos'), logpath, port, configdb, auth_param, extra)
+	      if os.name == 'nt':
```
Comment (kevinadi) [inline mtools/mlaunch/mlaunch.py]: There's a mix of tabs & spaces in this line

File: mtools/mlaunch/mlaunch.py [hunk 5]
```diff
@@ -1422,7 +1439,11 @@ def _construct_mongos(self, logpath, port, configdb):
...
+            newLogPath=logpath.replace('\\', '\\\\')
+            command_str = "start /b %s --logpath %s --port %i --configdb %s --logappend %s %s "%(os.path.join(path, 'mongos'), newLogPath, port, configdb, auth_param, extra)
+	      else:
```
Comment (kevinadi) [inline mtools/mlaunch/mlaunch.py]: There's a mix of tabs & spaces in this line

**Comment by kevinadi** [comment]
The Travis build was using Ubuntu 12.04.5 LTS. However, the failures were due to a mix of tabs & spaces in lines 1442 and 1445.

**Comment by zhaoyi0113** [comment]
Ok, thanks for pointing out. I have fixed the error and pushed. Could you let me know what else need to be done in order to merge in this PR?

**Comment by kevinadi** [comment]
@zhaoyi0113 thanks for the update. Sorry it took some time to review this.

I have found one last thing to update. Apparently extra parameters are not passed on to `mlaunch` on Windows. For example:

```
mlaunch init --single --storageEngine mmapv1
```

The `--storageEngine` parameter is not passed on in Windows, but is on OSX.

I tracked the issue to the `_filter_valid_arguments()` function, specifically in line 1062.

Do you mind updating the PR to accommodate extra parameters? Thanks.

File: mtools/mlaunch/mlaunch.py [hunk 6]
```diff
@@ -1128,7 +1130,13 @@ def _start_on_ports(self, ports, wait=False, overrideAuth=False):
                 command_str = re.sub(r'--keyFile \S+', '', command_str)
 
             try:
-                ret = subprocess.check_output([command_str], stderr=subprocess.STDOUT, shell=True)
+                if os.name == 'nt':
+                    # print "command str ", filter(None, command_str.split(' ')), "."
```
Comment (kevinadi) [inline mtools/mlaunch/mlaunch.py]: Could you remove this commented line please? I believe it's a development debug line, so we don't need it anymore. Thanks.

**Comment by kevinadi** [comment]
Also I have put up a comment regarding the development debug line (line 1134), I think we don't need that line anymore 👍

**Comment by zhaoyi0113** [comment]
I have fixed the `mlaunch init --single --storageEngine mmapv1` issue and removed the comment code. Please let me know what else need to be done for this PR. Thanks.

**Comment by stennie** [comment]
@zhaoyi0113 Great stuff .. thanks for your contribution!

Regards,
Stennie

**Pull request merged**